src/EventIndexing.js

```diff
@@ -0,0 +1,406 @@
+            // commit them now.
+            console.log("EventIndex: Committing events");
+            await indexManager.commitLiveEvents();
+            return;
+        }
+    }
+
+    async onRoomTimeline(ev, room, toStartOfTimeline, removed, data) {
+        const indexManager = PlatformPeg.get().getEventIndexingManager();
+        if (indexManager === null) return;
+
+        // We only index encrypted rooms locally.
+        if (!MatrixClientPeg.get().isRoomEncrypted(room.roomId)) return;
+
+        // If it isn't a live event or if it's redacted there's nothing to
+        // do.
+        if (toStartOfTimeline || !data || !data.liveEvent
+            || ev.isRedacted()) {
+            return;
+        }
+
+        // If the event is not yet decrypted mark it for the
+        // Event.decrypted callback.
+        if (ev.isBeingDecrypted()) {
+            const eventId = ev.getId();
+            this.liveEventsForIndex.add(eventId);
+        } else {
+            // If the event is decrypted or is unencrypted add it to the
+            // index now.
+            await this.addLiveEventToIndex(ev);
+        }
+    }
+
+    async onEventDecrypted(ev, err) {
+        const indexManager = PlatformPeg.get().getEventIndexingManager();
+        if (indexManager === null) return;
+
+        const eventId = ev.getId();
+
+        // If the event isn't in our live event set, ignore it.
+        if (!this.liveEventsForIndex.delete(eventId)) return;
+        if (err) return;
+        await this.addLiveEventToIndex(ev);
+    }
+
+    async addLiveEventToIndex(ev) {
+        const indexManager = PlatformPeg.get().getEventIndexingManager();
+        if (indexManager === null) return;
+
+        if (["m.room.message", "m.room.name", "m.room.topic"]
+            .indexOf(ev.getType()) == -1) {
+            return;
+        }
+
+        const e = ev.toJSON().decrypted;
+        const profile = {
+            displayname: ev.sender.rawDisplayName,
+            avatar_url: ev.sender.getMxcAvatarUrl(),
+        };
+
+        indexManager.addEventToIndex(e, profile);
+    }
+
+    async crawlerFunc(handle) {
+        // TODO either put this in a better place or find a library provided
+        // method that does this.
+        const sleep = async (ms) => {
+            return new Promise(resolve => setTimeout(resolve, ms));
+        };
+
+        let cancelled = false;
+
+        console.log("EventIndex: Started crawler function");
+
+        const client = MatrixClientPeg.get();
+        const indexManager = PlatformPeg.get().getEventIndexingManager();
+
+        handle.cancel = () => {
+            cancelled = true;
+        };
+
+        while (!cancelled) {
+            // This is a low priority task and we don't want to spam our
+            // Homeserver with /messages requests so we set a hefty timeout
+            // here.
+            await sleep(this._crawlerTimeout);
+
+            console.log("EventIndex: Running the crawler loop.");
```

Comment:
Any chance it can log some basic "M out of N" style progress?

Reply:
Depends on what you want the numbers to represent. The UI, which will come in a later PR, will have something like "Currently crawling N out of M encrypted rooms".

We do not know how far into the timeline we are and when we will reach an end.

Reply:
Okay, sounds like Matthew, Nad, and others will help advise on what's needed. Now that we have the feature flag, it feels safer to land with out it for now, since you have to opt-in.